### DISCUSSION TITLE
Implementation of webhook alerts

### DISCUSSION THREAD

**Discussion by henrygd** [henrygd/beszel#49] [Ideas]
### Update: This has been added in 0.4.0 using [Shoutrrr](https://containrrr.dev/shoutrrr/v0.8/services/overview/)

As with the additional disks thread (#36), this is a feature I plan to add at some point and want to source opinions on the best way to go about it.

I don't have enough free time to do custom implementations for every possible service. So I'm thinking something like the mock up below would work for the largest number of applications.

You would have `{{title}}` and `{{body}}` variables that would match the title and body of email alerts. This should allow you a bit of freedom in how you format the message. Including using JSON for the body.

I'd like to know if this absolutely won't work for anyone. And I'm open to suggestions. But again, I want to avoid writing custom code to support specific services.

![image](https://github.com/user-attachments/assets/4f80d620-adb1-4e69-8319-506fb2c33150)

**Comment by henrygd** [comment]
Also, this would be defined universally for each user. You wouldn't need to set it up for every alert. We could have a toggle for enabling / disabling. If it's enabled, it will send automatically for every alert.

**Comment by cichy1173** [comment]
Hello. Yes. Webhooks alerts will cover most of notification systems used in selfhosted world (as far as I see on social media) - for example, ntfy, Telegram bots or Home Assistant (maybe even Pagerduty if set in right way). That should cover most of use cases. Maybe somewhere in the future it would be good to add support for MQTT, but Webhooks will get job done. 

Right now only way to get notification about an alert is creating some grabber service, that will check for triggered alerts from frontend side, get alert data and do something with this which I see is supported thanks to PocketBase, but  it is not that straight forward like normal REST API with, ie curl calls.

**Comment by svenvg93** [comment]
I know from another project I worked on for notifications that depending on the service the payload might need to be diffrent. For example Discord expects "content" followed by the message, where ntfy expects "message" and Slack expexts "text" . Just my to cents.

**Reply by henrygd** [reply]
In the request body, right? In theory this approach should still work by putting the JSON content in the body field:

```json
{
    "topic": "mytopic",
    "message": "{{body}}",
    "title": "{{title}}",
    "tags": ["warning"],
    "priority": 4
}
```

**Reply by svenvg93** [reply]
Yess, If you look at discord it wants "content" for the body message instead message https://discord.com/developers/docs/resources/webhook#execute-webhook-jsonform-params 

Slack expects "test" for the body message 
https://api.slack.com/messaging/webhooks#posting_with_webhooks 

It won't be a one webhook to cover them all unfortunately :(

**Reply by henrygd** [reply]
I might be missing your point. For discord couldn't you just change the field name to "content"?

```json
{
    "content": "{{body}}"
}
```

**Reply by svenvg93** [reply]
Yess that what I meant ;). It was just a heads up :)

**Comment by Zottelchen** [comment]
Another way would be to integrate something like apprise (https://github.com/caronc/apprise). Not sure if there is a go library similar to it.
There seems to be PingMe (https://github.com/kha7iq/pingme), but the services seem rather limited.

**Reply by cw-cdz** [reply]
> Another way would be to integrate something like apprise (https://github.com/caronc/apprise). Not sure if there is a go library similar to it. There seems to be PingMe (https://github.com/kha7iq/pingme), but the services seem rather limited.

I agree. Integrating with apprise or shoutrrr or /similiar would provide an easy way to support many different providers, and those projects also have extensibility features (plugins)

**Comment by Tohjuler** [comment]
Good idea, you could hook up a template system.
Then people could contribute with templates, that way it could be a good fit for all users both them that is not as advance as others.

**Comment by parnic** [comment]
Feels like it would be simplest to integrate shoutrrr support and let users define what notifier to use.

**Reply by parnic** [reply]
https://github.com/henrygd/beszel/pull/145

**Comment by cichy1173** [comment]
Hi @henrygd . Thanks for adding sending alarms on webhook. I really like the feature. Also I see that I can add multiple webhooks, which is cool but I do not know how to take advantage of that. Is there a way to spread alarms across webhooks, for example, alarms from instance XYZ goes to webhook A, and alarms for instance ZYX goes to webhook B?

**Reply by henrygd** [reply]
Currently you would need to create a second user. Add them to system ZYX (in PocketBase - don't recreate it) and set up the alert with that user.

**Reply by cichy1173** [reply]
That's actually works. Thanks!

**Comment by FPille** [comment]
Signal webhook
`generic://localhost:8080/v2/send?template=json&disabletls=yes&$number=%2B491111111&$recipient=%2B492222222`

The credit goes to   dpmpc
https://github.com/containrrr/shoutrrr/issues/66#issuecomment-2541869735

**Reply by henrygd** [reply]
Nice, thanks! Good to know in case anyone asks.
